src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java

```diff
@@ -0,0 +1,342 @@
+            }
+            if (!combined) {
+                newRanges.add(selectedRange);
+            }
+        }
+        return newRanges;
+    }
+
+    /**
+     * Checks if the given condition collides with this condition.
+     *
+     * @param versionRangesRequestCondition The condition to check.
+     * @return True if the conditions collide, false otherwise.
+     */
+    public boolean collide(@NotNull VersionRangesRequestCondition versionRangesRequestCondition) {
+        Set<VersionRange> tbcRanges = versionRangesRequestCondition.getRanges();
+
+        // If one of those represents all versions, the other one has to collide
+        if (ranges.isEmpty() || tbcRanges.isEmpty()) {
+            return true;
+        }
+
+        for (VersionRange range : ranges) {
+            for (VersionRange tbcRange : tbcRanges) {
+                if (inRangeCodes.contains(VersionRangeComparator.compare(range, tbcRange))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public Set<VersionRange> getRanges() {
+        return ranges;
+    }
+
+    @Override
+    public int compareTo(@NotNull VersionRangesRequestCondition versionRangesRequestCondition, @NotNull HttpServletRequest httpServletRequest) {
+        if (this.getMatchingCondition(httpServletRequest) == this) {
+            return -1;
+        }
+        else if (versionRangesRequestCondition.getMatchingCondition(httpServletRequest) == versionRangesRequestCondition) {
+            return 1;
+        }
+        else {
+            return 0;
+        }
+    }
+
+    /**
+     * Returns the condition that matches the request or {@code null} if there is no match.
+     *
+     * @param request the current request
+     * @return the matching condition or {@code null}
+     */
+    @Nullable
+    @Override
+    public VersionRangesRequestCondition getMatchingCondition(@NotNull HttpServletRequest request) {
+        String path = request.getRequestURI();
+        if (!path.matches("^/?api/.*")) {
+            return null;
+        }
+        // If any version is valid, return this
+        if (ranges.isEmpty()) {
+            return this;
+        }
+
+        for (VersionRange range : ranges) {
+            if (versionRangeMatchesRequest(range, request)) {
+                return this;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Checks if the range matches the request.
+     *
+     * @param range the version range to match
+     * @param request the current request
+     * @return true if the range matches the request, false otherwise
+     */
+    public boolean versionRangeMatchesRequest(@NotNull VersionRange range, @NotNull HttpServletRequest request) {
+        String path = request.getRequestURI();
+        // Expect optional leading slash, "api/", optional numerical version with a "v" prefix and some path afterwards
+        var pattern = Pattern.compile("^/?api(/v(\\d*))?/.+$");
```

Comment:
```suggestion
        var pattern = Pattern.compile("^/?api(/v(\\d+))?/.+$");
```
Should `/api/v/…` be valid?

Reply:
Don't see how this would be valid. The `?` is on `(/v(\\d+))`

Reply:
I think it would be: for `/api/v/…` the `(/v(\\d*))?` the `?`-group part is there since the `/v` part exists, and `\\d*` allows optional digits.

See https://regexr.com/74mr8 first test: it is set to ‘Match none’ but the regex matches.

Reply:
Oh wow. That works differently than I expected. Good catch, then.

Seems like I won't really be able to do that with a regex completely. Probably need to strip it afterwards.

Reply:
Now I get it. Still didn't work as I expected but that doesn't hurt us here.
I completely didn't see your suggested change before.

Yes, that makes sense of course.

The fact that `/api/v/test` is working, is okay. It's just the path `/v/test` without a version.